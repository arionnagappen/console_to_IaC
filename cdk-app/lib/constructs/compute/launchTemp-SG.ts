import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';

export interface SgProps {
  vpc: ec2.IVpc
}

export class AppSecurityGroup extends Construct {
  public readonly securityGroup: ec2.SecurityGroup;

  constructor(scope: Construct, id: string, props: SgProps) {
    super(scope, id);

    this.securityGroup = new ec2.SecurityGroup(this, 'AppServerSG', {
      vpc: props.vpc,
      description: 'App Server Security Group',  
      allowAllOutbound: false
    });

    
    // Only allow inbound traffic on port 80
    this.securityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80)
    )

    // Only allow outbound traffic on port 80
    this.securityGroup.addEgressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80),
    )

  }
}