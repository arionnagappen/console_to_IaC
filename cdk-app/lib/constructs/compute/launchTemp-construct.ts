import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { ISecurityGroup } from "aws-cdk-lib/aws-ec2";

export interface LaunchTempProps {
  launchTempName: string;
  imageId: string;
  instanceType: string;
  securityGroupId: ISecurityGroup["securityGroupId"];
}

export class MyLaunchTemp extends Construct {
  public readonly launchTemp: ec2.CfnLaunchTemplate

  constructor(scope: Construct, id: string, props: LaunchTempProps) {
    super(scope, id)

    this.launchTemp = new ec2.CfnLaunchTemplate(this, 'AppLaunchTemplate', {
      launchTemplateName: props.launchTempName,
      launchTemplateData: {
        imageId: props.imageId, 
        instanceType: props.instanceType,
        securityGroupIds: [props.securityGroupId],
      }
    });
  };
};