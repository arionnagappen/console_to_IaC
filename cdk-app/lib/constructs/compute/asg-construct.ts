import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { MyLaunchTemp } from './launchTemp-construct';
import { CfnAutoScalingGroup } from 'aws-cdk-lib/aws-autoscaling';
import { ISubnet } from 'aws-cdk-lib/aws-ec2';

export interface AsgProps {
  vpc: ec2.IVpc;
  launchTemplate: ec2.CfnLaunchTemplate;
  subnetType?: ec2.SubnetType;
  minSize?: number;
  maxSize?: number;
  desiredCapacity?: number;
}

export class MyAutoScalingGroup extends Construct {
  public readonly asg: CfnAutoScalingGroup;

  constructor(scope: Construct, id: string, props: AsgProps) {
    super(scope, id);

    const subnets = props.vpc.selectSubnets({
      subnetType: props.subnetType?? ec2.SubnetType.PUBLIC
    })

    this.asg = new CfnAutoScalingGroup(this, 'AppServerASG', {
      minSize: `${props.minSize ?? 1}`,
      maxSize: `${props.maxSize ?? 3}`,
      desiredCapacity: `${props.desiredCapacity ?? 2}`,
      vpcZoneIdentifier: subnets.subnetIds,
      launchTemplate: {
        launchTemplateId: props.launchTemplate.ref,
        version: props.launchTemplate.attrLatestVersionNumber
      }
    });
  }
}