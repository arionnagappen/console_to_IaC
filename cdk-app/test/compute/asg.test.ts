import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { MyLaunchTemp } from '../../lib/constructs/compute/launchTemp-construct';
import { AppSecurityGroup } from '../../lib/constructs/compute/launchTemp-SG';
import { MyAutoScalingGroup } from '../../lib/constructs/compute/asg-construct';

// Assertion Test
test('Check Properties of Auto-Scaling Group', () => {
  const app = new cdk.App();

  const stack = new cdk.Stack(app, 'TestStack');

  const vpc = new ec2.Vpc(stack, 'TestVpc', {
    maxAzs: 1,
    subnetConfiguration: [{
      name: 'test-subnet',
      subnetType: ec2.SubnetType.PUBLIC
    }]
  });

  const sg = new AppSecurityGroup(stack, 'TestSG', { vpc });

  const launchTemplate = new MyLaunchTemp(stack, 'TestLaunchTemplate', {
    launchTempName: 'TestLaunchTemplate',
    imageId: 'ami-12345678',
    instanceType: 't3.micro',
    securityGroupId: sg.securityGroup.securityGroupId,
  });

  // Create ASG
  new MyAutoScalingGroup(stack, 'TestASG', {
    vpc,
    subnetType: ec2.SubnetType.PUBLIC,
    minSize: 1,
    maxSize: 3,
    desiredCapacity: 2,
    launchTemplate: launchTemplate.launchTemp
  });

  // Assert
  const template = Template.fromStack(stack);

  template.hasResourceProperties('AWS::AutoScaling::AutoScalingGroup', {
    MinSize: '1',
    MaxSize: '3',
    DesiredCapacity: '2'
  });
});

// Snapshot Test
test('ASG Snapshot Test', ()=> {
  const app = new cdk.App();

  const stack = new cdk.Stack(app, 'TestStack');

  const vpc = new ec2.Vpc(stack, 'TestVpc', {
    maxAzs: 1,
    subnetConfiguration: [{
      name: 'test-subnet',
      subnetType: ec2.SubnetType.PUBLIC
    }]
  });

  const sg = new AppSecurityGroup(stack, 'TestSG', { vpc });

  const launchTemplate = new MyLaunchTemp(stack, 'TestLaunchTemplate', {
    launchTempName: 'TestLaunchTemplate',
    imageId: 'ami-12345678',
    instanceType: 't3.micro',
    securityGroupId: sg.securityGroup.securityGroupId,
  });

  // Create ASG
  new MyAutoScalingGroup(stack, 'TestASG', {
    vpc,
    subnetType: ec2.SubnetType.PUBLIC,
    minSize: 1,
    maxSize: 3,
    desiredCapacity: 2,
    launchTemplate: launchTemplate.launchTemp
  });

  // Snapshot
  const template = Template.fromStack(stack).toJSON()
  expect(template).toMatchSnapshot();
})