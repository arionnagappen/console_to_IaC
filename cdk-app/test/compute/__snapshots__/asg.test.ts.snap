// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ASG Snapshot Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestASGAppServerASGE3140A3D": {
      "Properties": {
        "DesiredCapacity": "2",
        "LaunchTemplate": {
          "LaunchTemplateId": {
            "Ref": "TestLaunchTemplateAppLaunchTemplate816C81D4",
          },
          "Version": {
            "Fn::GetAtt": [
              "TestLaunchTemplateAppLaunchTemplate816C81D4",
              "LatestVersionNumber",
            ],
          },
        },
        "MaxSize": "3",
        "MinSize": "1",
        "VPCZoneIdentifier": [
          {
            "Ref": "TestVpctestsubnetSubnet1Subnet212A675B",
          },
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
    },
    "TestLaunchTemplateAppLaunchTemplate816C81D4": {
      "Properties": {
        "LaunchTemplateData": {
          "ImageId": "ami-12345678",
          "InstanceType": "t3.micro",
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "TestSGAppServerSG063DFCD3",
                "GroupId",
              ],
            },
          ],
        },
        "LaunchTemplateName": "TestLaunchTemplate",
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
    "TestSGAppServerSG063DFCD3": {
      "Properties": {
        "GroupDescription": "App Server Security Group",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestVpcE77CE678": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcIGW9DD53F70": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVpcVPCGWF1827B84": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "TestVpctestsubnetSubnet1DefaultRoute30ABFF53": {
      "DependsOn": [
        "TestVpcVPCGWF1827B84",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "RouteTableId": {
          "Ref": "TestVpctestsubnetSubnet1RouteTableA3F3FD85",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpctestsubnetSubnet1RouteTableA3F3FD85": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/test-subnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpctestsubnetSubnet1RouteTableAssociation9E8E56C2": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpctestsubnetSubnet1RouteTableA3F3FD85",
        },
        "SubnetId": {
          "Ref": "TestVpctestsubnetSubnet1Subnet212A675B",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpctestsubnetSubnet1Subnet212A675B": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/16",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "test-subnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/test-subnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
